diff --git a/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamBluray.h b/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamBluray.h
index fc7940bbed..ab79c5fce2 100644
--- a/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamBluray.h
+++ b/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamBluray.h
@@ -30,7 +30,8 @@ extern "C"
 #define BD_EVENT_MENU_ERROR   -2
 #define BD_EVENT_ENC_ERROR    -3
 
-#define HDMV_PID_VIDEO            0x1011
+#define HDMV_PID_VIDEO_FIRST      0x1011
+#define HDMV_PID_VIDEO_LAST       0x1015
 #define HDMV_PID_AUDIO_FIRST      0x1100
 #define HDMV_PID_AUDIO_LAST       0x111f
 #define HDMV_PID_PG_FIRST         0x1200
@@ -42,6 +43,8 @@ extern "C"
 
 class CDVDOverlayImage;
 class IVideoPlayer;
+class CDVDInputStreamBlurayExtension;
+class FFmpegExtraData;
 
 class CDVDInputStreamBluray
   : public CDVDInputStream
@@ -65,6 +68,13 @@ public:
   int GetBlockSize() override { return 6144; }
   ENextStream NextStream() override;
 
+  /* BD3D related */
+  int Get3dSubtitlePlane(uint16_t pid);
+  inline bool AreEyesFlipped() const { return m_flipEyes; }
+  DemuxPacket *ReadExtensionPacket();
+  inline bool IsExtensionStreamFound() const { return m_extension != nullptr; }
+  FFmpegExtraData GetExtensionExtraData();
+  void SetExtensionStreamStartTime(int64_t ms);
 
   /* IMenus */
   void ActivateButton() override { UserInput(BD_VK_ENTER); }
@@ -186,12 +196,18 @@ protected:
     bool OpenStream(CFileItem &item);
     void SetupPlayerSettings();
     void FreeTitleInfo();
+    void SelectPlaylist(int item);
+
     std::unique_ptr<CDVDInputStreamFile> m_pstream;
     std::string m_rootPath;
 
     /*! Bluray state serializer handler */
     CBlurayStateSerializer m_blurayStateSerializer;
 
+    /* BD3D related things */
+    CDVDInputStreamBlurayExtension     *m_extension;
+    bool m_flipEyes;
+
     /* used during bd_open_stream read block*/
     CCriticalSection m_readBlocksLock;
 };
