diff --git a/xbmc/guilib/StereoscopicsManager.cpp b/xbmc/guilib/StereoscopicsManager.cpp
index 3aa269bca3..e281b550a7 100644
--- a/xbmc/guilib/StereoscopicsManager.cpp
+++ b/xbmc/guilib/StereoscopicsManager.cpp
@@ -52,6 +52,10 @@ static const struct StereoModeMap VideoModeToGuiModeMap[] =
   { "right_left",               RENDER_STEREO_MODE_SPLIT_VERTICAL },
   { "top_bottom",               RENDER_STEREO_MODE_SPLIT_HORIZONTAL },
   { "bottom_top",               RENDER_STEREO_MODE_SPLIT_HORIZONTAL },
+  { "full_left_right",          RENDER_STEREO_MODE_HARDWAREBASED },
+  { "full_right_left",          RENDER_STEREO_MODE_HARDWAREBASED },
+  { "full_top_bottom",          RENDER_STEREO_MODE_HARDWAREBASED },
+  { "full_bottom_top",          RENDER_STEREO_MODE_HARDWAREBASED },
   { "checkerboard_rl",          RENDER_STEREO_MODE_CHECKERBOARD },
   { "checkerboard_lr",          RENDER_STEREO_MODE_CHECKERBOARD },
   { "row_interleaved_rl",       RENDER_STEREO_MODE_INTERLACED },
@@ -61,8 +65,8 @@ static const struct StereoModeMap VideoModeToGuiModeMap[] =
   { "anaglyph_cyan_red",        RENDER_STEREO_MODE_ANAGLYPH_RED_CYAN },
   { "anaglyph_green_magenta",   RENDER_STEREO_MODE_ANAGLYPH_GREEN_MAGENTA },
   { "anaglyph_yellow_blue",     RENDER_STEREO_MODE_ANAGLYPH_YELLOW_BLUE },
-  { "block_lr",                 RENDER_STEREO_MODE_OFF }, // unsupported
-  { "block_rl",                 RENDER_STEREO_MODE_OFF }, // unsupported
+  { "block_lr",                 RENDER_STEREO_MODE_HARDWAREBASED },
+  { "block_rl",                 RENDER_STEREO_MODE_HARDWAREBASED },
   {}
 };
 
@@ -130,10 +134,76 @@ void CStereoscopicsManager::SetStereoMode(const RENDER_STEREO_MODE &mode)
   RENDER_STEREO_MODE currentMode = GetStereoMode();
   RENDER_STEREO_MODE applyMode = mode;
 
+  const auto& components = CServiceBroker::GetAppComponents();
+  const auto appPlayer = components.GetComponent<CApplicationPlayer>();
+
   // resolve automatic mode before applying
   if (mode == RENDER_STEREO_MODE_AUTO)
     applyMode = GetStereoModeOfPlayingVideo();
 
+  // Check if we're allowed to switch the stereo mode
+  if (CServiceBroker::GetSettingsComponent()->GetSettings()->GetInt(CSettings::SETTING_VIDEOPLAYER_ADJUSTREFRESHRATE) == ADJUST_REFRESHRATE_OFF)
+  {
+    // mode switches are not allowed
+    if (currentMode == RENDER_STEREO_MODE_MONO || currentMode == RENDER_STEREO_MODE_OFF)
+    {
+      // we are in 2D mode
+      if (IsVideoStereoscopic())
+      {
+        // no mode switches allowed, but we've got a 3D video and are running in 2D then set the stereo mode to mono
+        applyMode = RENDER_STEREO_MODE_MONO;
+      }
+    }
+    else if (IsVideoStereoscopic())
+    {
+      // no mode switches allowed, and we've got a 3D video and are running in 3D => done
+      applyMode = currentMode;
+    }
+    else
+    {
+      // TODO: we're in 3D mode, the video is 2D and mode switching is not allowed,
+      // then we've a bit of bad luck because we can't do anything about it :-(
+      applyMode = currentMode;
+    }
+
+    CLog::Log(LOGDEBUG, "StereoscopicsManager: mode switching not allowed, using stereo mode {}", ConvertGuiStereoModeToString(applyMode));
+  }
+  else if (appPlayer->IsPlaying() && IsVideoStereoscopic())
+  {
+    // a 3D video is playing, let's check if we really support the requested stereoscopic mode
+    float fps = CServiceBroker::GetDataCacheCore().GetVideoFps();
+    int width = CServiceBroker::GetDataCacheCore().GetVideoWidth();
+    int height = CServiceBroker::GetDataCacheCore().GetVideoHeight();
+
+    // request a matching resolution
+    uint32_t mode3dFlags = CServiceBroker::GetWinSystem()->GetGfxContext().ConvertRenderStereoModeToMode3dFlags(applyMode);
+    RESOLUTION res = CResolutionUtils::ChooseBestResolution(static_cast<float>(fps), width, height, mode3dFlags);
+    RENDER_STEREO_MODE rsm = CServiceBroker::GetWinSystem()->GetGfxContext().GetRenderStereoModeFromResolution(res);
+
+    // check if we've got what we've requested
+    if (rsm != applyMode && rsm == RENDER_STEREO_MODE_OFF)
+    {
+      // nope, check for possible fallbacks in case of MVC/FP playback
+      if (applyMode == RENDER_STEREO_MODE_HARDWAREBASED)
+      {
+        // check SBS fallback for MVC/FP output
+   	    res = CResolutionUtils::ChooseBestResolution(static_cast<float>(fps), width, height, D3DPRESENTFLAG_MODE3DSBS);
+   	    rsm = CServiceBroker::GetWinSystem()->GetGfxContext().GetRenderStereoModeFromResolution(res);
+   	    if (rsm != RENDER_STEREO_MODE_SPLIT_VERTICAL)
+   	    {
+          // check TAB fallback for MVC/FP output
+   	      res = CResolutionUtils::ChooseBestResolution(static_cast<float>(fps), width, height, D3DPRESENTFLAG_MODE3DTB);
+   	      rsm = CServiceBroker::GetWinSystem()->GetGfxContext().GetRenderStereoModeFromResolution(res);
+          if (rsm != RENDER_STEREO_MODE_SPLIT_HORIZONTAL)
+        	rsm = RENDER_STEREO_MODE_MONO;
+   	    }
+      	applyMode = rsm;
+      }
+      else
+        applyMode = RENDER_STEREO_MODE_MONO;
+    }
+  }
+
   if (applyMode != currentMode && applyMode >= RENDER_STEREO_MODE_OFF)
   {
     if (CServiceBroker::GetRenderSystem()->SupportsStereo(applyMode))
@@ -198,6 +268,15 @@ std::string CStereoscopicsManager::DetectStereoModeByString(const std::string &n
   if (re.RegFind(searchString) > -1)
     stereoMode = "top_bottom";
 
+  if (!re.RegComp(CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_stereoscopicregex_mvc))
+  {
+    CLog::Log(LOGERROR, "{}: Invalid RegExp for matching 3d MVC content:'{}'", __FUNCTION__, CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_stereoscopicregex_mvc);
+    return stereoMode;
+  }
+
+  if (re.RegFind(searchString) > -1)
+    stereoMode = "block_lr";
+
   return stereoMode;
 }
 
@@ -304,7 +383,7 @@ int CStereoscopicsManager::ConvertVideoToGuiStereoMode(const std::string &mode)
   size_t i = 0;
   while (VideoModeToGuiModeMap[i].name)
   {
-    if (mode == VideoModeToGuiModeMap[i].name)
+    if (mode == VideoModeToGuiModeMap[i].name && CServiceBroker::GetRenderSystem()->SupportsStereo(VideoModeToGuiModeMap[i].mode))
       return VideoModeToGuiModeMap[i].mode;
     i++;
   }
@@ -512,8 +591,25 @@ std::string CStereoscopicsManager::GetVideoStereoMode() const
 
   const auto& components = CServiceBroker::GetAppComponents();
   const auto appPlayer = components.GetComponent<CApplicationPlayer>();
-  if (appPlayer->IsPlaying())
-    playerMode = CServiceBroker::GetDataCacheCore().GetVideoStereoMode();
+  if (appPlayer->IsPlaying()) {
+	  playerMode = CServiceBroker::GetDataCacheCore().GetVideoStereoMode();
+
+	  if (playerMode.find("left") != std::string::npos) {
+		  // check for Full-SBS
+		  int width = CServiceBroker::GetDataCacheCore().GetVideoWidth();
+
+		  if (width == 3840) {
+			  playerMode = "full_" + playerMode;
+		  }
+	  } else if (playerMode.find("bottom") != std::string::npos) {
+		  // check for Full-TAB
+		  int height = CServiceBroker::GetDataCacheCore().GetVideoHeight();
+
+		  if (height > 1080) {
+			  playerMode = "full_" + playerMode;
+		  }
+	  }
+  }
 
   return playerMode;
 }
