diff --git a/xbmc/cores/VideoPlayer/DVDDemuxers/CMakeLists.txt b/xbmc/cores/VideoPlayer/DVDDemuxers/CMakeLists.txt
old mode 100644
new mode 100755
index 48710f847c..06dc3c38f8
--- a/xbmc/cores/VideoPlayer/DVDDemuxers/CMakeLists.txt
+++ b/xbmc/cores/VideoPlayer/DVDDemuxers/CMakeLists.txt
@@ -7,6 +7,9 @@ set(SOURCES DemuxMultiSource.cpp
             DVDDemuxFFmpeg.cpp
             DVDDemuxUtils.cpp
             DVDDemuxVobsub.cpp
+            DVDStreamMuxer.cpp
+            DVDDoviStreamMuxer.cpp
+            DVDMvcStreamMuxer.cpp
             DVDFactoryDemuxer.cpp)
 
 set(HEADERS DemuxMultiSource.h
@@ -18,6 +21,9 @@ set(HEADERS DemuxMultiSource.h
             DVDDemuxFFmpeg.h
             DVDDemuxUtils.h
             DVDDemuxVobsub.h
+            DVDStreamMuxer.h
+            DVDDoviStreamMuxer.h
+            DVDMvcStreamMuxer.h
             DVDFactoryDemuxer.h)
 
 core_add_library(dvddemuxers)
