diff --git a/xbmc/windowing/amlogic/VideoSyncAML.h b/xbmc/windowing/amlogic/VideoSyncAML.h
new file mode 100644
index 0000000000..55b23d9927
--- /dev/null
+++ b/xbmc/windowing/amlogic/VideoSyncAML.h
@@ -0,0 +1,26 @@
+/*
+ *  Copyright (C) 2017-2018 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+#include "windowing/VideoSync.h"
+#include "guilib/DispResource.h"
+
+class CVideoSyncAML : public CVideoSync, IDispResource
+{
+public:
+  CVideoSyncAML(CVideoReferenceClock *clock);
+  virtual ~CVideoSyncAML();
+  virtual bool Setup()override;
+  virtual void Run(CEvent& stopEvent)override;
+  virtual void Cleanup()override;
+  virtual float GetFps()override;
+  virtual void OnResetDisplay()override;
+private:
+  volatile bool m_abort;
+};
