diff --git a/xbmc/windowing/WinSystem.cpp b/xbmc/windowing/WinSystem.cpp
index f1d40bf2f0..0ba9b62ba5 100644
--- a/xbmc/windowing/WinSystem.cpp
+++ b/xbmc/windowing/WinSystem.cpp
@@ -77,6 +77,8 @@ void CWinSystemBase::UpdateDesktopResolution(RESOLUTION_INFO& newRes,
     newRes.strMode += "i";
   if (dwFlags & D3DPRESENTFLAG_MODE3DTB)
     newRes.strMode += "tab";
+  if (dwFlags & D3DPRESENTFLAG_MODE3DFP)
+    newRes.strMode += "fp";
   if (dwFlags & D3DPRESENTFLAG_MODE3DSBS)
     newRes.strMode += "sbs";
   newRes.strOutput = output;
@@ -158,9 +160,9 @@ static bool resSortPredicate(const RESOLUTION_WHR& i, const RESOLUTION_WHR& j)
 {
   // note: this comparison must obey "strict weak ordering"
   // a "!=" on the flags comparison resulted in memory corruption
-  return (    i.width < j.width
-          || (i.width == j.width && i.height < j.height)
-          || (i.width == j.width && i.height == j.height && i.flags < j.flags) );
+  return (    i.m_screenWidth < j.m_screenWidth
+          || (i.m_screenWidth == j.m_screenWidth && i.m_screenHeight < j.m_screenHeight)
+          || (i.m_screenWidth == j.m_screenWidth && i.m_screenHeight == j.m_screenHeight && i.flags < j.flags) );
 }
 
 std::vector<RESOLUTION_WHR> CWinSystemBase::ScreenResolutions(float refreshrate)
