--- a/xbmc/settings/DisplaySettings.cpp
+++ b/xbmc/settings/DisplaySettings.cpp
@@ -76,6 +76,8 @@
     res += "sbs";
   else if(flags & D3DPRESENTFLAG_MODE3DTB)
     res += "tab";
+  else if(flags & D3DPRESENTFLAG_MODE3DFP)
+    res += "fp";
   else if(identifier)
     res += "std";
   return res;
@@ -253,7 +255,6 @@
 {
   if (!setting)
     return false;
-
   const std::string &settingId = setting->GetId();
   if (settingId == CSettings::SETTING_VIDEOSCREEN_RESOLUTION ||
       settingId == CSettings::SETTING_VIDEOSCREEN_SCREEN)
@@ -291,10 +292,21 @@
     // in which case we don't show any prompt to the user
     if (oldRes != RES_WINDOW && newRes != RES_WINDOW && oldRes != newRes)
     {
-      if (!m_resolutionChangeAborted)
-      {
-        if (HELPERS::ShowYesNoDialogText(CVariant{13110}, CVariant{13111}, CVariant{""},
-                                         CVariant{""}, 15000) != DialogResponse::CHOICE_YES)
+        if (!m_resolutionChangeAborted)
+        {
+          /* If new resolution is 4K, then warn user this is not advised */
+          RESOLUTION_INFO resInfo = GetResolutionInfo(newRes);
+          if (resInfo.iScreenHeight >= 2160)
+          {
+            CLog::Log(LOGWARNING, "The user has chosen a UHD resolution. We will warn them not to apply this");
+            if (HELPERS::ShowYesNoDialogText(CVariant{13110}, CVariant{38384}, CVariant{""}, CVariant{""}, 15000) != DialogResponse::CHOICE_YES)
+            {
+              m_resolutionChangeAborted = true;
+              return false;
+            }
+        }
+        if (HELPERS::ShowYesNoDialogText(CVariant{13110}, CVariant{13111}, CVariant{""}, CVariant{""}, 15000) !=
+          DialogResponse::CHOICE_YES)
         {
           m_resolutionChangeAborted = true;
           return false;
@@ -501,24 +513,6 @@
   std::unique_lock lock(m_critical);
   RESOLUTION_INFO res(resolution);
 
-  if((res.dwFlags & D3DPRESENTFLAG_MODE3DTB) == 0)
-  {
-    /* add corrections for some special case modes frame packing modes */
-
-    if(res.iScreenWidth  == 1920
-    && res.iScreenHeight == 2205)
-    {
-      res.iBlanking = 45;
-      res.dwFlags  |= D3DPRESENTFLAG_MODE3DTB;
-    }
-
-    if(res.iScreenWidth  == 1280
-    && res.iScreenHeight == 1470)
-    {
-      res.iBlanking = 30;
-      res.dwFlags  |= D3DPRESENTFLAG_MODE3DTB;
-    }
-  }
   m_resolutions.push_back(res);
 }
 
@@ -531,7 +525,14 @@
     // find resolutions
     for (size_t res = RES_DESKTOP; res < m_resolutions.size(); ++res)
     {
-      if (StringUtils::EqualsNoCase(itCal->strMode, m_resolutions[res].strMode))
+      if (StringUtils::EqualsNoCase(itCal->strMode, m_resolutions[res].strMode) && (m_resolutions[res].dwFlags & D3DPRESENTFLAG_MODE3DMASK))
+      {
+        // in case of 3D resolutions we ignore overscan settings and just update the subtitle position
+        m_resolutions[res].iSubtitles = itCal->iSubtitles;
+        if (m_resolutions[res].iSubtitles < 0)
+          m_resolutions[res].iSubtitles = 0;
+      }
+      else if (StringUtils::EqualsNoCase(itCal->strMode, m_resolutions[res].strMode))
       {
         // overscan
         m_resolutions[res].Overscan.left = itCal->Overscan.left;
@@ -569,7 +570,6 @@
           m_resolutions[res].fPixelRatio = 0.5f;
         if (m_resolutions[res].fPixelRatio > 2.0f)
           m_resolutions[res].fPixelRatio = 2.0f;
-        break;
       }
     }
   }
@@ -600,6 +600,8 @@
       cal = *res;
     }
   }
+
+  ApplyCalibrations();
 }
 
 void CDisplaySettings::ClearCalibrations()
@@ -669,6 +671,8 @@
       flags |= D3DPRESENTFLAG_MODE3DSBS;
     else if(StringUtils::Mid(strResolution, 20,3) == "tab")
       flags |= D3DPRESENTFLAG_MODE3DTB;
+    else if(StringUtils::Mid(strResolution, 20, 2) == "fp")
+      flags |= D3DPRESENTFLAG_MODE3DFP;
 
     std::map<RESOLUTION, RESOLUTION_INFO> resolutionInfos;
     for (size_t resolution = RES_DESKTOP; resolution < CDisplaySettings::GetInstance().ResolutionInfoSize(); resolution++)
@@ -681,6 +685,37 @@
   return RES_DESKTOP;
 }
 
+std::string CDisplaySettings::GetDisplayStringFromRes(const RESOLUTION res, bool hiDpi, bool withRefreshRate) const
+{
+  RESOLUTION_INFO info = GetResolutionInfo(res);
+
+  std::string stereo = "(2D)";
+
+  if (info.dwFlags & D3DPRESENTFLAG_MODE3DMASK)
+  {
+    stereo = "(3D ";
+    if (info.dwFlags & D3DPRESENTFLAG_MODE3DSBS)
+      stereo += "SBS)";
+    else if(info.dwFlags & D3DPRESENTFLAG_MODE3DTB)
+      stereo += "TAB)";
+    else if((info.dwFlags & D3DPRESENTFLAG_MODE3DFP) && (info.dwFlags & D3DPRESENTFLAG_MODE3DFSBS))
+      stereo += "Full-SBS)";
+    else if((info.dwFlags & D3DPRESENTFLAG_MODE3DFP) && (info.dwFlags & D3DPRESENTFLAG_MODE3DFTAB))
+      stereo += "Full-TAB)";
+    else if(info.dwFlags & D3DPRESENTFLAG_MODE3DFP)
+      stereo += "FP)";
+  }
+
+  std::string refRate = "";
+  if (withRefreshRate) {
+	  refRate = StringUtils::Format("{:0.2f} Hz ", info.fRefreshRate);
+  }
+
+  return StringUtils::Format("{}x{}{} {}{}{}",
+      info.iScreenWidth, info.iScreenHeight, info.dwFlags & D3DPRESENTFLAG_INTERLACED ? 'i' : 'p',
+      refRate, stereo, hiDpi ? " (HiDPI)" : "");
+}
+
 std::string CDisplaySettings::GetStringFromResolution(RESOLUTION resolution, float refreshrate /* = 0.0f */)
 {
   if (resolution == RES_WINDOW)
@@ -690,13 +725,9 @@
       resolution < static_cast<RESOLUTION>(CDisplaySettings::GetInstance().ResolutionInfoSize()))
   {
     const RESOLUTION_INFO &info = CDisplaySettings::GetInstance().GetResolutionInfo(resolution);
-    // also handle RES_DESKTOP resolutions with non-default refresh rates
-    if (resolution != RES_DESKTOP || (refreshrate > 0.0f && refreshrate != info.fRefreshRate))
-    {
-      return StringUtils::Format("{:05}{:05}{:09.5f}{}", info.iScreenWidth, info.iScreenHeight,
-                                 refreshrate > 0.0f ? refreshrate : info.fRefreshRate,
-                                 ModeFlagsToString(info.dwFlags, true));
-    }
+    return StringUtils::Format("{:05}{:05}{:09.5f}{}",
+        info.iScreenWidth, info.iScreenHeight,
+        refreshrate > 0.0f ? refreshrate : info.fRefreshRate, ModeFlagsToString(info.dwFlags, true));
   }
 
   return "DESKTOP";
@@ -721,15 +752,8 @@
     const RESOLUTION_INFO& mode = CDisplaySettings::GetInstance().GetResolutionInfo(index);
 
     if (mode.dwFlags ^ D3DPRESENTFLAG_INTERLACED)
-    {
-      const std::string screenmode =
-          GetStringFromResolution(static_cast<RESOLUTION>(index), mode.fRefreshRate);
-
-      list.emplace_back(
-          StringUtils::Format("{}x{}{} {:0.2f}Hz", mode.iScreenWidth, mode.iScreenHeight,
-                              ModeFlagsToString(mode.dwFlags, false), mode.fRefreshRate),
-          screenmode);
-    }
+      list.emplace_back(CDisplaySettings::GetInstance().GetDisplayStringFromRes(static_cast<RESOLUTION>(index)),
+                        GetStringFromResolution(static_cast<RESOLUTION>(index), mode.fRefreshRate));
   }
 
   std::ranges::sort(list, [](const StringSettingOption& i, const StringSettingOption& j)
@@ -801,14 +825,8 @@
     std::vector<RESOLUTION_WHR> resolutions = CServiceBroker::GetWinSystem()->ScreenResolutions(info.fRefreshRate);
     for (const auto& resolution : resolutions)
     {
-      const std::string resLabel =
-          StringUtils::Format("{}x{}{}{}", resolution.m_screenWidth, resolution.m_screenHeight,
-                              ModeFlagsToString(resolution.flags, false),
-                              resolution.width > resolution.m_screenWidth &&
-                                      resolution.height > resolution.m_screenHeight
-                                  ? " (HiDPI)"
-                                  : "");
-      list.emplace_back(resLabel, resolution.ResInfo_Index);
+      bool hiDpi = resolution.width > resolution.m_screenWidth && resolution.height > resolution.m_screenHeight;
+      list.emplace_back(CDisplaySettings::GetInstance().GetDisplayStringFromRes(static_cast<RESOLUTION>(resolution.ResInfo_Index), hiDpi, false), resolution.ResInfo_Index);
 
       resolutionInfos.try_emplace(
           static_cast<RESOLUTION>(resolution.ResInfo_Index),
