diff --git a/xbmc/windowing/amlogic/WinSystemAmlogic.cpp b/xbmc/windowing/amlogic/WinSystemAmlogic.cpp
new file mode 100755
index 0000000000..b65b0fe3e4
--- /dev/null
+++ b/xbmc/windowing/amlogic/WinSystemAmlogic.cpp
@@ -0,0 +1,511 @@
+/*
+ *  Copyright (C) 2005-2025 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "WinSystemAmlogic.h"
+
+#include <string>
+
+#include "ServiceBroker.h"
+#include "cores/AudioEngine/AESinkFactory.h"
+#include "cores/AudioEngine/Sinks/AESinkALSA.h"
+#include "cores/RetroPlayer/process/amlogic/RPProcessInfoAmlogic.h"
+#include "cores/RetroPlayer/rendering/VideoRenderers/RPRendererOpenGLES.h"
+#include "cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAmlogic.h"
+#include "cores/VideoPlayer/VideoRenderers/LinuxRendererGLES.h"
+#include "cores/VideoPlayer/VideoRenderers/HwDecRender/RendererAML.h"
+#include "messaging/ApplicationMessenger.h"
+#include "platform/linux/ScreenshotSurfaceAML.h"
+#include "settings/DisplaySettings.h"
+#include "settings/Settings.h"
+#include "settings/VeroSettings.h"
+#include "settings/SettingsComponent.h"
+#include "settings/SettingPath.h"
+#include "utils/log.h"
+#include "utils/AMLUtils.h"
+#include "utils/StringUtils.h"
+#include "utils/SysfsUtils.h"
+#include "windowing/GraphicContext.h"
+#include "windowing/Resolution.h"
+
+
+using namespace KODI;
+using namespace KODI::MESSAGING;
+
+
+CWinSystemAmlogic::CWinSystemAmlogic()
+    : m_libinput(new CLibInputHandler)
+{
+  const char *env_framebuffer = getenv("FRAMEBUFFER");
+
+  // default to framebuffer 0
+  m_framebuffer_name = "fb0";
+  if (env_framebuffer) {
+    std::string framebuffer(env_framebuffer);
+    std::string::size_type start = framebuffer.find("fb");
+    m_framebuffer_name = framebuffer.substr(start);
+  }
+
+  m_displayWidth = 0;
+  m_displayHeight = 0;
+
+  m_stereo_mode = RENDER_STEREO_MODE_OFF;
+  m_delayDispReset = false;
+
+  aml_permissions();
+  aml_disable_freeScale();
+
+  m_settings = CServiceBroker::GetSettingsComponent()->GetSettings();
+
+  /* Take in to account custom OSMC parameters */
+  m_veroSettings = new VeroSettings();
+
+  SysfsUtils::GetString("/sys/class/amhdmitx/amhdmitx0/rawedid", m_lastEdid);
+
+  StartMonitorHWEvent();
+
+  // Register sink
+  AE::CAESinkFactory::ClearSinks();
+  CAESinkALSA::Register();
+  m_libinput->Start();
+}
+
+CWinSystemAmlogic::~CWinSystemAmlogic()
+{
+  StopMonitorHWEvent();
+  delete m_veroSettings;
+}
+
+void CWinSystemAmlogic::hwMonitorThread()
+{
+  struct udev *udev;
+  struct udev_device *dev;
+
+  struct udev_monitor *mon;
+  int fd;
+
+  /* Create the udev object */
+  udev = udev_new();
+  if (!udev) {
+    return;
+  }
+
+  mon = udev_monitor_new_from_netlink(udev, "udev");
+  udev_monitor_filter_add_match_subsystem_devtype(mon, "extcon", NULL);
+  udev_monitor_filter_add_match_subsystem_devtype(mon, "switch", NULL);
+
+  udev_monitor_enable_receiving(mon);
+  fd = udev_monitor_get_fd(mon);
+
+  /* Poll for events */
+  while (m_monitorEvents) {
+    fd_set fds;
+    struct timeval tv;
+    int ret;
+
+    FD_ZERO(&fds);
+    FD_SET(fd, &fds);
+    tv.tv_sec = 0;
+    tv.tv_usec = 0;
+
+    ret = select(fd + 1, &fds, NULL, NULL, &tv);
+
+    /* Check if FD has received data */
+
+    if (ret > 0 && FD_ISSET(fd, &fds)) {
+      dev = udev_monitor_receive_device(mon);
+      if (dev) {
+        CLog::Log(LOGDEBUG, "CWinSystemAmlogic: Detected HDMI switch");
+
+        int state;
+        SysfsUtils::GetInt("/sys/class/amhdmitx/amhdmitx0/hpd_state", state);
+
+        std::string newEdid;
+        SysfsUtils::GetString("/sys/class/amhdmitx/amhdmitx0/rawedid", newEdid);
+
+        if (state && newEdid != m_lastEdid) {
+          CServiceBroker::GetAppMessenger()->PostMsg(TMSG_AML_RESIZE);
+          m_lastEdid = newEdid;
+        }
+        udev_device_unref(dev);
+      } else {
+        CLog::Log(LOGERROR, "CWinSystemAmlogic: can't get device from receive_device");
+      }
+    }
+    usleep(250 * 1000);
+  }
+}
+
+void CWinSystemAmlogic::StartMonitorHWEvent()
+{
+  CLog::Log(LOGDEBUG, "CWinSystemAmlogic::StartMonitorHWEvent -- starting event monitor for HDMI hotplug events");
+  m_monitorEvents = true;
+  m_monitorThread = std::thread(&CWinSystemAmlogic::hwMonitorThread, this);
+}
+
+void CWinSystemAmlogic::StopMonitorHWEvent()
+{
+  CLog::Log(LOGDEBUG, "CWinSystemAmlogic::StopMonitorHWEvent -- stopping event monitor for HDMI hotplug events");
+  m_monitorEvents = false;
+  m_monitorThread.join();
+}
+
+bool CWinSystemAmlogic::InitWindowSystem()
+{
+  CDVDVideoCodecAmlogic::Register();
+  CLinuxRendererGLES::Register();
+  RETRO::CRPProcessInfoAmlogic::Register();
+  RETRO::CRPProcessInfoAmlogic::RegisterRendererFactory(new RETRO::CRendererFactoryOpenGLES);
+  CRendererAML::Register();
+  CScreenshotSurfaceAML::Register();
+
+  aml_set_framebuffer_resolution(4096, 2205, m_framebuffer_name);
+
+  return CWinSystemBase::InitWindowSystem();
+}
+
+bool CWinSystemAmlogic::DestroyWindowSystem()
+{
+  m_bWindowCreated = false;
+
+  return true;
+}
+
+bool CWinSystemAmlogic::CreateNewWindow(const std::string &name, bool fullScreen, RESOLUTION_INFO &res)
+{
+  RESOLUTION_INFO current_resolution;
+  RENDER_STEREO_MODE stereo_mode = CServiceBroker::GetWinSystem()->GetGfxContext().GetStereoMode();
+  bool dvOutput = IsDolbyVisionOutput();
+
+  if (m_bWindowCreated && aml_get_native_resolution(&current_resolution)
+      && current_resolution.iWidth == res.iWidth && current_resolution.iHeight == res.iHeight
+      && current_resolution.iScreenWidth == res.iScreenWidth && current_resolution.iScreenHeight == res.iScreenHeight
+      && m_bFullScreen == fullScreen && current_resolution.fRefreshRate == res.fRefreshRate
+      && (current_resolution.dwFlags & D3DPRESENTFLAG_MODEMASK) == (res.dwFlags & D3DPRESENTFLAG_MODEMASK)
+      && m_stereo_mode == stereo_mode && m_dvOutput == dvOutput) {
+    CLog::Log(LOGDEBUG, "CWinSystemAmlogic::CreateNewWindow: No need to create a new window");
+    return true;
+  }
+
+  m_nWidth = res.iWidth;
+  m_nHeight = res.iHeight;
+  m_displayWidth = res.iScreenWidth;
+  m_displayHeight = res.iScreenHeight;
+  m_fRefreshRate = res.fRefreshRate;
+  m_stereo_mode = stereo_mode;
+  m_bFullScreen = fullScreen;
+  m_dvOutput = dvOutput;
+
+  OnLostDisplay();
+
+  if (!aml_set_native_resolution(res, m_framebuffer_name, stereo_mode)) {
+    CLog::Log(LOGERROR, "CWinSystemAmlogic::CreateNewWindow -- unable to set native resolution");
+    return false;
+  }
+
+  ResetDisplay();
+
+  m_bWindowCreated = true;
+
+  return true;
+}
+
+void CWinSystemAmlogic::ResetDisplay()
+{
+  if (!IsDisplayResetDue()) {
+    int delay = m_settings->GetInt("videoscreen.delayrefreshchange");
+    if (delay > 0) {
+      m_delayDispReset = true;
+      m_dispResetTimer.Set(std::chrono::milliseconds(static_cast<unsigned int>(delay * 100)));
+      return;
+    }
+  }
+
+  std::unique_lock<CCriticalSection> lock(m_resourceSection);
+
+  // tell any shared resources
+  for (IDispResource *resource : m_resources) {
+    resource->OnResetDisplay();
+  }
+
+  m_delayDispReset = false;
+}
+
+void CWinSystemAmlogic::OnLostDisplay()
+{
+  std::unique_lock<CCriticalSection> lock(m_resourceSection);
+
+  for (IDispResource *resource : m_resources) {
+    resource->OnLostDisplay();
+  }
+}
+
+bool CWinSystemAmlogic::IsDisplayResetDue() const
+{
+  return m_delayDispReset && m_dispResetTimer.IsTimePast();
+}
+
+bool CWinSystemAmlogic::DestroyWindow()
+{
+  return true;
+}
+
+static std::string ModeFlagsToString(unsigned int flags, bool identifier)
+{
+  std::string res;
+
+  if (flags & D3DPRESENTFLAG_INTERLACED) {
+    res += "i";
+  } else {
+    res += "p";
+  }
+
+  if (!identifier) {
+    res += " ";
+  }
+
+  if (flags & D3DPRESENTFLAG_MODE3DSBS) {
+    res += "sbs";
+  } else if (flags & D3DPRESENTFLAG_MODE3DTB) {
+    res += "tab";
+  } else if (flags & D3DPRESENTFLAG_MODE3DFP) {
+    res += "fp";
+  } else if (identifier) {
+    res += "std";
+  }
+
+  return res;
+}
+
+void CWinSystemAmlogic::Update3dResolutions()
+{
+  if (!aml_probe_3d_resolutions(resolutions3d) || resolutions3d.empty()) {
+    CLog::Log(LOGWARNING, "{}: No valid 3D resolutions found.", __FUNCTION__);
+    return;
+  }
+
+  for (size_t i = 0; i < resolutions3d.size(); i++) {
+    if (resolutions3d[i].dwFlags & D3DPRESENTFLAG_INTERLACED) { // ignore interlaced 3D resolutions
+      continue;
+    }
+
+    std::string m3d = "<unknown>";
+    if (resolutions3d[i].dwFlags & D3DPRESENTFLAG_MODE3DSBS) {
+      m3d = "SBS";
+    } else if (resolutions3d[i].dwFlags & D3DPRESENTFLAG_MODE3DTB) {
+      m3d = "TAB";
+    } else if (resolutions3d[i].dwFlags & D3DPRESENTFLAG_MODE3DFP) {
+      m3d = "FP";
+    }
+
+    CServiceBroker::GetWinSystem()->GetGfxContext().ResetOverscan(resolutions3d[i]);
+    CDisplaySettings::GetInstance().AddResolutionInfo(resolutions3d[i]);
+
+    CLog::Log(LOGINFO, "Found 3D resolution {} x {} with {} x {}{} @ {:f} ({})",
+              resolutions3d[i].iWidth, resolutions3d[i].iHeight,
+              resolutions3d[i].iScreenWidth, resolutions3d[i].iScreenHeight,
+              resolutions3d[i].dwFlags & D3DPRESENTFLAG_INTERLACED ? "i" : "",
+              resolutions3d[i].fRefreshRate, m3d);
+  }
+}
+
+void CWinSystemAmlogic::UpdateResolutions()
+{
+  CWinSystemBase::UpdateResolutions();
+
+  RESOLUTION_INFO resDesktop, curDisplay;
+  std::string curDesktopSetting, curResolution, newRes;
+  std::vector<RESOLUTION_INFO> resolutions;
+
+  if (!aml_probe_resolutions(resolutions) || resolutions.empty()) {
+    CLog::Log(LOGWARNING, "{}: ProbeResolutions failed.", __FUNCTION__);
+  }
+
+  /* ProbeResolutions includes already all resolutions.
+   * Only get desktop resolution so we can replace xbmc's desktop res
+   */
+  if (aml_get_native_resolution(&curDisplay)) {
+    resDesktop = curDisplay;
+  }
+
+  curDesktopSetting = m_settings->GetString(CSettings::SETTING_VIDEOSCREEN_SCREENMODE);
+
+  curResolution = StringUtils::Format("{:05}{:05}{:09.5f}{}", resDesktop.iScreenWidth, resDesktop.iScreenHeight,
+                                      resDesktop.fRefreshRate, ModeFlagsToString(resDesktop.dwFlags, true));
+
+  if (curDesktopSetting == "DESKTOP") {
+    curDesktopSetting = curResolution;
+  } else if (curDesktopSetting.length() == 24) {
+    curDesktopSetting = StringUtils::Right(curDesktopSetting, 23);
+  }
+
+  CLog::Log(LOGINFO, "Current display setting is {}", curDesktopSetting);
+  CLog::Log(LOGINFO, "Current output resolution is {}", curResolution);
+
+  RESOLUTION ResDesktop = RES_INVALID;
+  RESOLUTION res_index = RES_DESKTOP;
+  bool resExactMatch = false;
+  std::string ResString;
+  std::string ResFallback = "00480024.00000istd";
+
+ for (size_t i = 0; i < resolutions.size(); i++) {
+    // if this is a new setting,
+    // create a new empty setting to fill in.
+    if ((int) CDisplaySettings::GetInstance().ResolutionInfoSize() <= res_index) {
+      RESOLUTION_INFO res;
+      CDisplaySettings::GetInstance().AddResolutionInfo(res);
+    }
+
+    CServiceBroker::GetWinSystem()->GetGfxContext().ResetOverscan(resolutions[i]);
+    CDisplaySettings::GetInstance().GetResolutionInfo(res_index) = resolutions[i];
+
+    CLog::Log(LOGINFO, "Found resolution {:d} x {:d} with {:d} x {:d}{} @ {:f} Hz",
+              resolutions[i].iWidth, resolutions[i].iHeight,
+              resolutions[i].iScreenWidth, resolutions[i].iScreenHeight,
+              resolutions[i].dwFlags & D3DPRESENTFLAG_INTERLACED ? "i" : "",
+              resolutions[i].fRefreshRate);
+
+    ResString = StringUtils::Format("{:05}{:05}{:09.5f}{}",
+                                    resolutions[i].iScreenWidth, resolutions[i].iScreenHeight, resolutions[i].fRefreshRate,
+                                    ModeFlagsToString(resolutions[i].dwFlags, true));
+
+    if (curDesktopSetting == ResString) {
+      ResDesktop = res_index;
+      resExactMatch = true;
+      newRes = ResString;
+      CLog::Log(LOGINFO, "Current resolution setting found at 16 + {}", i);
+    }
+
+    /* fall back to the highest resolution available but not more than current desktop */
+    if (curDesktopSetting.substr(5, 18).compare(ResString.substr(5, 18)) >= 0 && ResString.substr(5, 18).compare(ResFallback) > 0 && !resExactMatch) {
+      ResDesktop = res_index;
+      ResFallback = ResString.substr(5, 18);
+      newRes = ResString;
+      CLog::Log(LOGINFO, "Fallback resolution at 16 + {} {}", i, ResFallback);
+    }
+
+    res_index = (RESOLUTION) ((int) res_index + 1);
+  }
+
+  // set RES_DESKTOP
+  if (ResDesktop != RES_INVALID) {
+    CLog::Log(LOGINFO, "Found {} at {:d}, setting to RES_DESKTOP at {:d}", newRes, (int) ResDesktop, (int) RES_DESKTOP);
+    CDisplaySettings::GetInstance().GetResolutionInfo(RES_DESKTOP) = CDisplaySettings::GetInstance().GetResolutionInfo(ResDesktop);
+  }
+
+  Update3dResolutions();
+  CDisplaySettings::GetInstance().ApplyCalibrations();
+}
+
+bool CWinSystemAmlogic::Hide()
+{
+  return false;
+}
+
+bool CWinSystemAmlogic::Show(bool show)
+{
+  std::string blank_framebuffer = "/sys/class/graphics/" + m_framebuffer_name + "/blank";
+
+  SysfsUtils::SetInt(blank_framebuffer.c_str(), show ? 0 : 1);
+
+  return true;
+}
+
+void CWinSystemAmlogic::Register(IDispResource *resource)
+{
+  std::unique_lock<CCriticalSection> lock(m_resourceSection);
+  m_resources.push_back(resource);
+}
+
+void CWinSystemAmlogic::Unregister(IDispResource *resource)
+{
+  std::unique_lock<CCriticalSection> lock(m_resourceSection);
+
+  std::vector<IDispResource*>::iterator i = find(m_resources.begin(), m_resources.end(), resource);
+  if (i != m_resources.end()) {
+    m_resources.erase(i);
+  }
+}
+
+CHDRCapabilities CWinSystemAmlogic::GetDisplayHDRCapabilities() const
+{
+  std::string hdr_caps;
+  CHDRCapabilities HDRcaps;
+
+  if (SysfsUtils::GetString("/sys/class/amvecm/hdr_support", hdr_caps) == 0) {
+    if (hdr_caps.find("HDR10") != std::string::npos) {
+      HDRcaps.SetHDR10();
+    }
+    if (hdr_caps.find("HDR10+") != std::string::npos) {
+      HDRcaps.SetHDR10Plus();
+    }
+    if (hdr_caps.find("HLG") != std::string::npos) {
+      HDRcaps.SetHLG();
+    }
+    if (hdr_caps.find("STD") != std::string::npos) {
+      HDRcaps.SetDolbyVision();
+    }
+    if (hdr_caps.find("LL") != std::string::npos) {
+      HDRcaps.SetDolbyVisionLL();
+    }
+  }
+
+  return HDRcaps;
+}
+
+bool CWinSystemAmlogic::IsHDRDisplay()
+{
+  return GetDisplayHDRCapabilities().SupportsHDR10();
+}
+
+float CWinSystemAmlogic::GetGuiSdrPeakLuminance() const
+{
+  const int guiSdrPeak = m_settings->GetInt(CSettings::SETTING_VIDEOSCREEN_GUISDRPEAKLUMINANCE);
+
+  return 0.5f * guiSdrPeak / 100.0f + 0.5f;
+}
+
+bool CWinSystemAmlogic::IsDolbyVisionSupportedByDisplay() const
+{
+  std::string dv_cap;
+
+  if (SysfsUtils::GetString("/sys/devices/virtual/amhdmitx/amhdmitx0/dv_cap", dv_cap)) {
+    return false;
+  }
+
+  return dv_cap.find("DV_RGB_444_8BIT") != std::string::npos;
+}
+
+bool CWinSystemAmlogic::IsDolbyVisionOutput() const
+{
+  if (!IsDolbyVisionSupportedByDisplay()) {
+    return false;
+  }
+
+  if (m_settings->GetInt(VeroSettings::SETTING_VIDEOPLAYER_HDR_PROCESS) != 2) {
+    return false;
+  }
+
+  std::string dv_enabled;
+
+  if (SysfsUtils::GetString("/sys/module/amdolby_vision/parameters/dolby_vision_enable", dv_enabled)) {
+    // unable to access sysfs
+    return false;
+  }
+
+  return dv_enabled != "N";
+}
+
+float CWinSystemAmlogic::GetDisplayLatency()
+{
+  if (m_fRefreshRate == 0.) {
+    // default display latency
+    return CWinSystemBase::GetDisplayLatency();
+  }
+
+  return 1000. / m_fRefreshRate * (CServiceBroker::GetWinSystem()->GetGfxContext().IsTransferPQ() ? 4. : 3.);
+}
