diff --git a/xbmc/cores/VideoPlayer/DVDCodecs/Video/amlogic/namespace.h b/xbmc/cores/VideoPlayer/DVDCodecs/Video/amlogic/namespace.h
new file mode 100644
index 0000000000..98f20775cc
--- /dev/null
+++ b/xbmc/cores/VideoPlayer/DVDCodecs/Video/amlogic/namespace.h
@@ -0,0 +1,32 @@
+/*
+ *  Copyright (C) 2005-2022 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+namespace amlogic {
+	// base classes
+	class AMLCodecList;
+	class AMLDecoderInputQueueStats;
+	class AMLInsecureVideoCodec;
+	class AMLVideoCodec;
+	class AMLVideoCodecInfo;
+	class DllLibAmCodec;
+	class LibAmcodec;
+	class PosixFile;
+
+	class DVCodec;
+
+	// insecure decoders
+	class AV1Codec;
+	class H264Codec;
+	class HEVCCodec;
+	class MPEG12Codec;
+	class MPEG4Codec;
+	class VC1Codec;
+	class VP9Codec;
+}
