diff --git a/xbmc/guilib/StereoscopicsManager.cpp b/xbmc/guilib/StereoscopicsManager.cpp
--- a/xbmc/guilib/StereoscopicsManager.cpp
+++ b/xbmc/guilib/StereoscopicsManager.cpp
@@ -53,6 +53,10 @@
     {"right_left", RenderStereoMode::SPLIT_VERTICAL},
     {"top_bottom", RenderStereoMode::SPLIT_HORIZONTAL},
     {"bottom_top", RenderStereoMode::SPLIT_HORIZONTAL},
+    {"full_left_right", RenderStereoMode::HARDWAREBASED},
+    {"full_right_left", RenderStereoMode::HARDWAREBASED},
+    {"full_top_bottom", RenderStereoMode::HARDWAREBASED},
+    {"full_bottom_top", RenderStereoMode::HARDWAREBASED},
     {"checkerboard_rl", RenderStereoMode::CHECKERBOARD},
     {"checkerboard_lr", RenderStereoMode::CHECKERBOARD},
     {"row_interleaved_rl", RenderStereoMode::INTERLACED},
@@ -62,8 +66,8 @@
     {"anaglyph_cyan_red", RenderStereoMode::ANAGLYPH_RED_CYAN},
     {"anaglyph_green_magenta", RenderStereoMode::ANAGLYPH_GREEN_MAGENTA},
     {"anaglyph_yellow_blue", RenderStereoMode::ANAGLYPH_YELLOW_BLUE},
-    {"block_lr", RenderStereoMode::OFF}, // unsupported
-    {"block_rl", RenderStereoMode::OFF}, // unsupported
+    {"block_lr", RenderStereoMode::HARDWAREBASED},
+    {"block_rl", RenderStereoMode::HARDWAREBASED},
     {}};
 
 static const struct StereoModeMap StringToGuiModeMap[] = {
@@ -127,10 +131,71 @@
   RenderStereoMode currentMode = GetStereoMode();
   RenderStereoMode applyMode = mode;
 
+  const auto& components = CServiceBroker::GetAppComponents();
+  const auto appPlayer = components.GetComponent<CApplicationPlayer>();
+
   // resolve automatic mode before applying
   if (mode == RenderStereoMode::AUTO)
     applyMode = GetStereoModeOfPlayingVideo();
 
+  // Check if we are allowed to switch the stereo mode
+  if (CServiceBroker::GetSettingsComponent()->GetSettings()->GetInt(CSettings::SETTING_VIDEOPLAYER_ADJUSTREFRESHRATE) == ADJUST_REFRESHRATE_OFF)
+  {
+    // mode switches are not allowed
+    if (currentMode == RenderStereoMode::MONO || currentMode == RenderStereoMode::OFF)
+    {
+      // we are in 2D mode
+      if (IsVideoStereoscopic())
+      {
+        // no mode switches allowed, but we have a 3D video and are running in 2D then set the stereo mode to mono
+        applyMode = RenderStereoMode::MONO;
+      }
+    }
+    else if (IsVideoStereoscopic())
+    {
+      // no mode switches allowed, and we have a 3D video and are running in 3D => done
+      applyMode = currentMode;
+    }
+    else
+    {
+      // TODO: we are in 3D mode, the video is 2D and mode switching is not allowed,
+      // then we have a bit of bad luck because we cannot do anything about it.
+      applyMode = currentMode;
+    }
+
+    CLog::Log(LOGDEBUG, "StereoscopicsManager: mode switching not allowed, using stereo mode {}", ConvertGuiStereoModeToString(applyMode));
+  }
+  else if (appPlayer->IsPlaying() && IsVideoStereoscopic())
+  {
+    // a 3D video is playing, check if we support the requested stereoscopic mode
+    float fps = CServiceBroker::GetDataCacheCore().GetVideoFps();
+    int width = CServiceBroker::GetDataCacheCore().GetVideoWidth();
+    int height = CServiceBroker::GetDataCacheCore().GetVideoHeight();
+
+    uint32_t mode3dFlags = CServiceBroker::GetWinSystem()->GetGfxContext().ConvertRenderStereoModeToMode3dFlags(applyMode);
+    RESOLUTION res = CResolutionUtils::ChooseBestResolution(static_cast<float>(fps), width, height, mode3dFlags);
+    RenderStereoMode rsm = CServiceBroker::GetWinSystem()->GetGfxContext().GetRenderStereoModeFromResolution(res);
+
+    if (rsm != applyMode && rsm == RenderStereoMode::OFF)
+    {
+      if (applyMode == RenderStereoMode::HARDWAREBASED)
+      {
+        res = CResolutionUtils::ChooseBestResolution(static_cast<float>(fps), width, height, D3DPRESENTFLAG_MODE3DSBS);
+        rsm = CServiceBroker::GetWinSystem()->GetGfxContext().GetRenderStereoModeFromResolution(res);
+        if (rsm != RenderStereoMode::SPLIT_VERTICAL)
+        {
+          res = CResolutionUtils::ChooseBestResolution(static_cast<float>(fps), width, height, D3DPRESENTFLAG_MODE3DTB);
+          rsm = CServiceBroker::GetWinSystem()->GetGfxContext().GetRenderStereoModeFromResolution(res);
+          if (rsm != RenderStereoMode::SPLIT_HORIZONTAL)
+            rsm = RenderStereoMode::MONO;
+        }
+        applyMode = rsm;
+      }
+      else
+        applyMode = RenderStereoMode::MONO;
+    }
+  }
+
   if (applyMode != currentMode && applyMode >= RenderStereoMode::OFF)
   {
     if (CServiceBroker::GetRenderSystem()->SupportsStereo(applyMode))
@@ -198,6 +263,15 @@
   if (re.RegFind(searchString) > -1)
     stereoMode = "top_bottom";
 
+  if (!re.RegComp(CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_stereoscopicregex_mvc.c_str()))
+  {
+    CLog::Log(LOGERROR, "{}: Invalid RegExp for matching 3d MVC content:'{}'", __FUNCTION__, CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_stereoscopicregex_mvc);
+    return stereoMode;
+  }
+
+  if (re.RegFind(searchString) > -1)
+    stereoMode = "block_lr";
+
   return stereoMode;
 }
 
@@ -307,7 +381,7 @@
   size_t i = 0;
   while (VideoModeToGuiModeMap[i].name)
   {
-    if (mode == VideoModeToGuiModeMap[i].name)
+    if (mode == VideoModeToGuiModeMap[i].name && CServiceBroker::GetRenderSystem()->SupportsStereo(VideoModeToGuiModeMap[i].mode))
       return VideoModeToGuiModeMap[i].mode;
     ++i;
   }
@@ -522,8 +596,25 @@
 
   const auto& components = CServiceBroker::GetAppComponents();
   const auto appPlayer = components.GetComponent<CApplicationPlayer>();
-  if (appPlayer->IsPlaying())
-    playerMode = CServiceBroker::GetDataCacheCore().GetVideoStereoMode();
+  if (appPlayer->IsPlaying())
+  {
+    playerMode = CServiceBroker::GetDataCacheCore().GetVideoStereoMode();
+
+    if (playerMode.find("left") != std::string::npos)
+    {
+      int width = CServiceBroker::GetDataCacheCore().GetVideoWidth();
+      if (width == 3840)
+        playerMode = "full_" + playerMode;
+    }
+    else if (playerMode.find("bottom") != std::string::npos)
+    {
+      int height = CServiceBroker::GetDataCacheCore().GetVideoHeight();
+      if (height > 1080)
+        playerMode = "full_" + playerMode;
+    }
+  }
 
   return playerMode;
 }
