diff --git a/xbmc/video/PlayerController.cpp b/xbmc/video/PlayerController.cpp
index 35025c517b..fa8604682b 100644
--- a/xbmc/video/PlayerController.cpp
+++ b/xbmc/video/PlayerController.cpp
@@ -502,8 +502,7 @@ bool CPlayerController::OnAction(const CAction &action)
           for (const CVariant &mode : indexList)
           {
             auto res = CDisplaySettings::GetInstance().GetResFromString(mode.asString());
-            const RESOLUTION_INFO info = CServiceBroker::GetWinSystem()->GetGfxContext().GetResInfo(res);
-            dialog->Add(info.strMode);
+            dialog->Add(CDisplaySettings::GetInstance().GetDisplayStringFromRes(res));
             if (res == currentRes)
               current = idx;
             idx++;
