diff --git a/xbmc/utils/BitstreamConverter.cpp b/xbmc/utils/BitstreamConverter.cpp
--- a/xbmc/utils/BitstreamConverter.cpp
+++ b/xbmc/utils/BitstreamConverter.cpp
@@ -435,7 +435,7 @@
       // valid hvcC data (bitstream) always starts with the value 1 (version)
       if (m_to_annexb)
       {
-        /**
+        /** @todo from Amlogic
         * It seems the extradata is encoded as hvcC format.
         * Temporarily, we support configurationVersion==0 until 14496-15 3rd
         * is finalized. When finalized, configurationVersion will be 1 and we
@@ -539,7 +539,7 @@
         if (m_convert_bitstream)
         {
           // convert demuxer packet from bitstream to bytestream (AnnexB)
-          int bytestream_size = 0;
+          uint32_t bytestream_size = 0;
           uint8_t* bytestream_buff = NULL;
 
           BitstreamConvert(demuxer_content, demuxer_bytes, &bytestream_buff, &bytestream_size);
@@ -690,6 +690,7 @@
   uint16_t unit_size;
   uint32_t total_size = 0;
   uint8_t *out = NULL, unit_nb, sps_done = 0, sps_seen = 0, pps_seen = 0;
+  uint8_t mvc_done = 0;
   const uint8_t* extradata = (uint8_t*)in_extradata + 4;
   static const uint8_t nalu_header[4] = {0, 0, 0, 1};
 
@@ -738,6 +739,18 @@
       if (unit_nb)
         pps_seen = 1;
     }
+
+    if (!unit_nb && !mvc_done++)
+    {
+      if (in_extrasize - total_size > 14 && memcmp(extradata + 8, "mvcC", 4) == 0)
+      {
+        // start over; take SPS and PPS from the mvcC atom
+        extradata += 12 + 5; // skip over mvcC atom header
+        unit_nb = *extradata++ & 0x1f;  // number of sps unit(s)
+        sps_done = 0;
+        pps_seen = 0;
+      }
+    }
   }
 
   if (out)
@@ -873,7 +886,7 @@
 bool CBitstreamConverter::BitstreamConvert(uint8_t* pData,
                                            int iSize,
                                            uint8_t** poutbuf,
-                                           int* poutbuf_size)
+                                           uint32_t* poutbuf_size)
 {
   // based on h264_mp4toannexb_bsf.c (ffmpeg)
   // which is Copyright (c) 2007 Benoit Fouet <benoit.fouet@free.fr>
@@ -883,7 +896,7 @@
   uint8_t* buf = pData;
   uint32_t buf_size = iSize;
   uint8_t unit_type, nal_sps, nal_pps, nal_sei;
-  int32_t nal_size;
+  uint32_t nal_size;
   uint32_t cumul_size = 0;
   const uint8_t* buf_end = buf + buf_size;
 
@@ -927,7 +940,7 @@
       unit_type = (*buf >> 1) & 0x3f;
     }
 
-    if (buf + nal_size > buf_end || nal_size <= 0)
+    if (nal_size > (buf_end - buf) || nal_size == 0)
       goto fail;
 
     // Don't add sps/pps if the unit already contain them
@@ -1032,7 +1045,7 @@
 }
 
 void CBitstreamConverter::BitstreamAllocAndCopy(uint8_t** poutbuf,
-                                                int* poutbuf_size,
+                                                uint32_t* poutbuf_size,
                                                 const uint8_t* sps_pps,
                                                 uint32_t sps_pps_size,
                                                 const uint8_t* in,
