diff --git a/xbmc/platform/linux/storage/UDevProvider.cpp b/xbmc/platform/linux/storage/UDevProvider.cpp
index 9e9f7f1..c8d9abc 100644
--- a/xbmc/platform/linux/storage/UDevProvider.cpp
+++ b/xbmc/platform/linux/storage/UDevProvider.cpp
@@ -144,6 +144,20 @@ void CUDevProvider::GetDisks(std::vector<CMediaSource>& disks, bool removable)
       continue;
     }
 
+    // filter out osmc boot partition
+    if (strcmp(mountpoint, "/boot") == 0)
+    {
+      udev_device_unref(device);
+      continue;
+    }
+
+    // filter out tee partition
+    if (strcmp(mountpoint, "/tee") == 0)
+    {
+      udev_device_unref(device);
+      continue;
+    }
+
     // filter out things mounted on /tmp
     if (strstr(mountpoint, "/tmp"))
     {
