diff --git a/xbmc/messaging/ApplicationMessenger.h b/xbmc/messaging/ApplicationMessenger.h
index 72cb0a8..058b740 100644
--- a/xbmc/messaging/ApplicationMessenger.h
+++ b/xbmc/messaging/ApplicationMessenger.h
@@ -94,6 +94,9 @@
 
+#define TMSG_SETOSMCWALKTHROUGHSTATE      TMSG_MASK_APPLICATION + 40
+#define TMSG_AML_RESIZE                   TMSG_MASK_APPLICATION + 41
+
 #define TMSG_GUI_INFOLABEL                TMSG_MASK_GUIINFOMANAGER + 0
 #define TMSG_GUI_INFOBOOL                 TMSG_MASK_GUIINFOMANAGER + 1
 #define TMSG_UPDATE_CURRENT_ITEM          TMSG_MASK_GUIINFOMANAGER + 2
@@ -100,6 +103,7 @@
 #define TMSG_CECTOGGLESTATE               TMSG_MASK_PERIPHERALS + 1
 #define TMSG_CECACTIVATESOURCE            TMSG_MASK_PERIPHERALS + 2
 #define TMSG_CECSTANDBY                   TMSG_MASK_PERIPHERALS + 3
+#define TMSG_CECACTIVATESOURCEOSMCWAKEUP  TMSG_MASK_PERIPHERALS + 4
 
 #define TMSG_GUI_DIALOG_OPEN              TMSG_MASK_WINDOWMANAGER + 1
 #define TMSG_GUI_ACTIVATE_WINDOW          TMSG_MASK_WINDOWMANAGER + 2
