diff --git a/xbmc/messaging/ApplicationMessenger.h b/xbmc/messaging/ApplicationMessenger.h
index 72cb0a8..058b740 100644
--- a/xbmc/messaging/ApplicationMessenger.h
+++ b/xbmc/messaging/ApplicationMessenger.h
@@ -90,3 +90,5 @@
 #define TMSG_SET_VOLUME                   TMSG_MASK_APPLICATION + 36
 #define TMSG_SET_MUTE                     TMSG_MASK_APPLICATION + 37
 #define TMSG_RESUMEAPP                    TMSG_MASK_APPLICATION + 38
+#define TMSG_SETOSMCWALKTHROUGHSTATE      TMSG_MASK_APPLICATION + 40
+#define TMSG_AML_RESIZE                   TMSG_MASK_APPLICATION + 41
@@ -100,6 +102,7 @@
 #define TMSG_CECTOGGLESTATE               TMSG_MASK_PERIPHERALS + 1
 #define TMSG_CECACTIVATESOURCE            TMSG_MASK_PERIPHERALS + 2
 #define TMSG_CECSTANDBY                   TMSG_MASK_PERIPHERALS + 3
+#define TMSG_CECACTIVATESOURCEOSMCWAKEUP  TMSG_MASK_PERIPHERALS + 4
 
 #define TMSG_GUI_DIALOG_OPEN              TMSG_MASK_WINDOWMANAGER + 1
 #define TMSG_GUI_ACTIVATE_WINDOW          TMSG_MASK_WINDOWMANAGER + 2
