diff --git a/xbmc/platform/linux/threads/ThreadImplLinux.cpp b/xbmc/platform/linux/threads/ThreadImplLinux.cpp
index 8c7edbb591..77a06bd431 100644
--- a/xbmc/platform/linux/threads/ThreadImplLinux.cpp
+++ b/xbmc/platform/linux/threads/ThreadImplLinux.cpp
@@ -28,11 +28,11 @@ namespace
 {
 
 constexpr auto nativeThreadPriorityMap = make_map<ThreadPriority, int>({
-    {ThreadPriority::LOWEST, -1},
-    {ThreadPriority::BELOW_NORMAL, -1},
-    {ThreadPriority::NORMAL, 0},
-    {ThreadPriority::ABOVE_NORMAL, 1},
-    {ThreadPriority::HIGHEST, 1},
+    {ThreadPriority::LOWEST, -3},
+    {ThreadPriority::BELOW_NORMAL, -3},
+    {ThreadPriority::NORMAL, -2},
+    {ThreadPriority::ABOVE_NORMAL, -1},
+    {ThreadPriority::HIGHEST, -1},
 });
 
 static_assert(static_cast<size_t>(ThreadPriority::PRIORITY_COUNT) == nativeThreadPriorityMap.size(),
