diff --git a/xbmc/application/ApplicationSkinHandling.cpp b/xbmc/application/ApplicationSkinHandling.cpp
index 5b91bb0..a853627 100644
--- a/xbmc/application/ApplicationSkinHandling.cpp
+++ b/xbmc/application/ApplicationSkinHandling.cpp
@@ -8,6 +8,8 @@
 
 #include "ApplicationSkinHandling.h"
 
+#include "Application.h"
+
 #include "FileItem.h"
 #include "FileItemList.h"
 #include "GUIInfoManager.h"
@@ -522,8 +524,14 @@ bool CApplicationSkinHandling::OnSettingChanged(const CSetting& setting)
     {
       // now we can finally reload skins
       std::string builtin("ReloadSkin");
-      if (settingId == CSettings::SETTING_LOOKANDFEEL_SKIN && m_confirmSkinChange)
+      if (settingId == CSettings::SETTING_LOOKANDFEEL_SKIN && m_confirmSkinChange &&
+          g_application.m_eOSMCWalkthroughState != g_application.OSMC_WALKTHROUGH_ISRUNNING)
         builtin += "(confirm)";
+
+      // Always make sure ISDONE is set or our changes won't progress. We hit this point if the user has selected a different skin in the OSMC setup wizard.
+      if (g_application.m_eOSMCWalkthroughState == g_application.OSMC_WALKTHROUGH_ISRUNNING)
+        g_application.SetOSMCWalkthroughState(g_application.OSMC_WALKTHROUGH_ISDONE);
+
       CServiceBroker::GetAppMessenger()->PostMsg(TMSG_EXECUTE_BUILT_IN, -1, -1, nullptr, builtin);
     }
   }
