diff --git a/xbmc/application/Application.h b/xbmc/application/Application.h
index 8f6d71c..81590a5 100644
--- a/xbmc/application/Application.h
+++ b/xbmc/application/Application.h
@@ -187,6 +187,11 @@ public:
 
   std::unique_ptr<CServiceManager> m_ServiceManager;
 
+  /*! \Allow OSMC's sigterm handler to know when to do the business or not, otherwise we get code on EXIT and POWERDOWN/REBOOT paths
+  */
+  bool m_ShuttingDown{false};
+  bool isShuttingDown() const { return m_ShuttingDown; }
+
   /*!
   \brief Locks calls from outside kodi (e.g. python) until framemove is processed.
   */
diff --git a/xbmc/application/ApplicationMessageHandling.cpp b/xbmc/application/ApplicationMessageHandling.cpp
index 5396c24..39900a6 100644
--- a/xbmc/application/ApplicationMessageHandling.cpp
+++ b/xbmc/application/ApplicationMessageHandling.cpp
@@ -118,7 +118,10 @@ void CApplicationMessageHandling::OnApplicationMessage(MESSAGING::ThreadMessage*
   {
     case TMSG_POWERDOWN:
       if (m_app.Stop(EXITCODE_POWERDOWN))
+      {
+        m_app.m_ShuttingDown = true;
         CServiceBroker::GetPowerManager().Powerdown();
+      }
       break;
 
     case TMSG_QUIT:
@@ -144,7 +147,10 @@ void CApplicationMessageHandling::OnApplicationMessage(MESSAGING::ThreadMessage*
     case TMSG_RESTART:
     case TMSG_RESET:
       if (m_app.Stop(EXITCODE_REBOOT))
+      {
+        m_app.m_ShuttingDown = true;
         CServiceBroker::GetPowerManager().Reboot();
+      }
       break;
 
     case TMSG_RESTARTAPP:
