diff --git a/xbmc/application/ApplicationMessageHandling.cpp b/xbmc/application/ApplicationMessageHandling.cpp
index 5dfc324c2e..0a7f81b5f8 100644
--- a/xbmc/application/ApplicationMessageHandling.cpp
+++ b/xbmc/application/ApplicationMessageHandling.cpp
@@ -217,6 +217,23 @@ void CApplicationMessageHandling::OnApplicationMessage(MESSAGING::ThreadMessage*
     }
     break;
 
+    case TMSG_SETOSMCWALKTHROUGHSTATE:
+    {
+      switch (pMsg->param1)
+      {
+        case 0:
+          m_app.SetOSMCWalkthroughState(m_app.OSMC_WALKTHROUGH_NOTRUNNING);
+          break;
+        case 1:
+          m_app.SetOSMCWalkthroughState(m_app.OSMC_WALKTHROUGH_ISRUNNING);
+          break;
+        case 2:
+          m_app.SetOSMCWalkthroughState(m_app.OSMC_WALKTHROUGH_ISDONE);
+          break;
+      }
+      break;
+    }
+
     case TMSG_NETWORKMESSAGE:
       m_app.m_ServiceManager->GetNetwork().NetworkMessage(
           static_cast<CNetworkBase::EMESSAGE>(pMsg->param1), pMsg->param2);
diff --git a/xbmc/messaging/ApplicationMessenger.h b/xbmc/messaging/ApplicationMessenger.h
index 6f8b05def9..4e3e4cf4a5 100644
--- a/xbmc/messaging/ApplicationMessenger.h
+++ b/xbmc/messaging/ApplicationMessenger.h
@@ -92,6 +92,7 @@
 #define TMSG_SET_MUTE                     TMSG_MASK_APPLICATION + 37
 #define TMSG_RESUMEAPP                    TMSG_MASK_APPLICATION + 38
 #define TMSG_PROCESS_DELETE_AFTER_WATCH   TMSG_MASK_APPLICATION + 39
+#define TMSG_SETOSMCWALKTHROUGHSTATE      TMSG_MASK_APPLICATION + 40
 
 #define TMSG_GUI_INFOLABEL                TMSG_MASK_GUIINFOMANAGER + 0
 #define TMSG_GUI_INFOBOOL                 TMSG_MASK_GUIINFOMANAGER + 1
