From 48785ed6fe15a4d2f7d41d34b773827171349ced Mon Sep 17 00:00:00 2001
From: "Chris \\\"koying\\\" Browet" <cbro@semperpax.com>
Date: Wed, 27 Apr 2016 18:44:34 +0200
Subject: [PATCH] FIX: save skin setting changes immediately

---
 xbmc/settings/SkinSettings.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/xbmc/settings/SkinSettings.cpp b/xbmc/settings/SkinSettings.cpp
index 00fa6aa8ef..8a85e2be5d 100644
--- a/xbmc/settings/SkinSettings.cpp
+++ b/xbmc/settings/SkinSettings.cpp
@@ -56,6 +56,7 @@ void CSkinSettings::SetString(int setting, const std::string& label) const
   if (!skin)
     return;
   skin->SetString(setting, label);
+  skin->SaveSettings();
 }
 
 int CSkinSettings::TranslateBool(const std::string& setting) const
@@ -84,6 +85,7 @@ void CSkinSettings::SetBool(int setting, bool set) const
   if (!skin)
     return;
   skin->SetBool(setting, set);
+  skin->SaveSettings();
 }
 
 void CSkinSettings::Reset(const std::string& setting) const
@@ -92,6 +94,7 @@ void CSkinSettings::Reset(const std::string& setting) const
   if (!skin)
     return;
   skin->Reset(setting);
+  skin->SaveSettings();
 }
 
 std::set<ADDON::CSkinSettingPtr> CSkinSettings::GetSettings() const
@@ -122,6 +125,7 @@ void CSkinSettings::Reset() const
     return;
 
   skin->Reset();
+  skin->SaveSettings();
 
   CGUIInfoManager& infoMgr = CServiceBroker::GetGUI()->GetInfoManager();
   infoMgr.ResetCache();
-- 
2.34.1

