diff --git a/xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp b/xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp
index 51bbbc514a..17224fd135 100644
--- a/xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp
+++ b/xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp
@@ -26,6 +26,8 @@
 #include "utils/log.h"
 #include "windowing/GraphicContext.h"
 #include "windowing/WinSystem.h"
+#include "guilib/GUIComponent.h"
+#include "guilib/StereoscopicsManager.h"
 
 #include <memory>
 #include <mutex>
@@ -680,8 +682,12 @@ RESOLUTION CRenderManager::GetResolution()
     return res;
 
   if (CServiceBroker::GetSettingsComponent()->GetSettings()->GetInt(CSettings::SETTING_VIDEOPLAYER_ADJUSTREFRESHRATE) != ADJUST_REFRESHRATE_OFF)
-    res = CResolutionUtils::ChooseBestResolution(m_fps, m_picture.iWidth, m_picture.iHeight,
-                                                 !m_picture.stereoMode.empty());
+
+  {
+    RENDER_STEREO_MODE rsm = CServiceBroker::GetGUI()->GetStereoscopicsManager().GetStereoMode();
+    uint32_t mode3dFlags = CServiceBroker::GetWinSystem()->GetGfxContext().ConvertRenderStereoModeToMode3dFlags(rsm);
+    res = CResolutionUtils::ChooseBestResolution(m_fps, m_picture.iWidth, m_picture.iHeight, mode3dFlags);
+  }
 
   return res;
 }
@@ -884,8 +890,10 @@ void CRenderManager::UpdateResolution()
     {
       if (CServiceBroker::GetSettingsComponent()->GetSettings()->GetInt(CSettings::SETTING_VIDEOPLAYER_ADJUSTREFRESHRATE) != ADJUST_REFRESHRATE_OFF && m_fps > 0.0f)
       {
-        RESOLUTION res = CResolutionUtils::ChooseBestResolution(
-            m_fps, m_picture.iWidth, m_picture.iHeight, !m_picture.stereoMode.empty());
+        RENDER_STEREO_MODE rsm = CServiceBroker::GetGUI()->GetStereoscopicsManager().GetStereoMode();
+        uint32_t mode3dFlags = CServiceBroker::GetWinSystem()->GetGfxContext().ConvertRenderStereoModeToMode3dFlags(rsm);
+        RESOLUTION res = CResolutionUtils::ChooseBestResolution(m_fps, m_picture.iWidth, m_picture.iHeight, mode3dFlags);
+
         CServiceBroker::GetWinSystem()->GetGfxContext().SetVideoResolution(res, false);
         UpdateLatencyTweak();
         if (m_pRenderer)
@@ -1164,10 +1172,9 @@ void CRenderManager::PrepareNextRender()
   }
 
   CLog::LogFC(LOGDEBUG, LOGAVTIMING,
-              "frameOnScreen: {:f} renderPts: {:f} nextFramePts: {:f} -> diff: {:f}  render: {} "
-              "forceNext: {}",
-              frameOnScreen, renderPts, nextFramePts, (renderPts - nextFramePts),
-              renderPts >= nextFramePts, m_forceNext);
+            "clock: {:.3f}, renderPts: {:.3f}, nextFramePts: {:.3f}, diff: {:.3f}, queue: {}, late: {}, force next: {}",
+            frameOnScreen / 1000000.0, renderPts / 1000000.0, nextFramePts / 1000000.0, (nextFramePts - renderPts) / 1000000.0,
+            m_queued.size(), m_lateframes, m_forceNext);
 
   bool combined = false;
   if (m_presentsourcePast >= 0)
@@ -1179,41 +1186,26 @@ void CRenderManager::PrepareNextRender()
 
   if (renderPts >= nextFramePts || m_forceNext)
   {
-    // see if any future queued frames are already due
-    auto iter = m_queued.begin();
-    int idx = *iter;
-    ++iter;
-    while (iter != m_queued.end())
-    {
-      // the slot for rendering in time is [pts .. (pts +  x * frametime)]
-      // renderer/drivers have internal queues, being slightly late here does not mean that
-      // we are really late. The likelihood that we recover decreases the greater m_lateframes
-      // get. Skipping a frame is easier than having decoder dropping one (lateframes > 10)
-      double x = (m_lateframes <= 6) ? 0.98 : 0;
-      if (renderPts < m_Queue[*iter].pts + x * frametime)
-        break;
-      idx = *iter;
-      ++iter;
-    }
+    double lateframes = (renderPts - nextFramePts) / frametime;
 
-    // skip late frames
-    while (m_queued.front() != idx)
+    while (lateframes >= 1.5 && m_queued.size() > 2)
     {
+      // skip late frames if possible; if the queue is almost empty, we don't skip
+      // even if we should to avoid emptying the queue too fast
       if (m_presentsourcePast >= 0)
       {
         m_discard.push_back(m_presentsourcePast);
-        m_QueueSkip++;
       }
       m_presentsourcePast = m_queued.front();
       m_queued.pop_front();
+      m_QueueSkip++;
+
+      lateframes = (renderPts - m_Queue[m_queued.front()].pts) / frametime;
     }
 
-    int lateframes = static_cast<int>((renderPts - m_Queue[idx].pts) *
-                                      static_cast<double>(m_fps / DVD_TIME_BASE));
-    if (lateframes)
-      m_lateframes += lateframes;
-    else
-      m_lateframes = 0;
+    int idx = m_queued.front();
+
+    m_lateframes = (int) (lateframes > 0. ? lateframes : 0.);
 
     m_presentstep = PRESENT_FLIP;
     m_discard.push_back(m_presentsource);
