diff --git a/xbmc/windowing/GraphicContext.h b/xbmc/windowing/GraphicContext.h
--- a/xbmc/windowing/GraphicContext.h
+++ b/xbmc/windowing/GraphicContext.h
@@ -24,11 +24,19 @@
 #define D3DPRESENTFLAG_PROGRESSIVE  4
 #define D3DPRESENTFLAG_MODE3DSBS    8
 #define D3DPRESENTFLAG_MODE3DTB    16
+#define D3DPRESENTFLAG_MODE3DFP    32
+#define D3DPRESENTFLAG_MODE3DFSBS  0x4000000
+#define D3DPRESENTFLAG_MODE3DFTAB  0x8000000
 
 /* what types are important for mode setting */
 #define D3DPRESENTFLAG_MODEMASK ( D3DPRESENTFLAG_INTERLACED \
                                   | D3DPRESENTFLAG_MODE3DSBS  \
-                                  | D3DPRESENTFLAG_MODE3DTB   )
+                                  | D3DPRESENTFLAG_MODE3DTB   \
+                                  | D3DPRESENTFLAG_MODE3DFP   )
+/* 3D modes only */
+#define D3DPRESENTFLAG_MODE3DMASK (D3DPRESENTFLAG_MODE3DSBS \
+                                   | D3DPRESENTFLAG_MODE3DTB \
+                                   | D3DPRESENTFLAG_MODE3DFP)
 
 enum VIEW_TYPE { VIEW_TYPE_NONE = 0,
                  VIEW_TYPE_LIST,
@@ -84,7 +92,6 @@
 
   void Flip(bool rendered, bool videoLayer);
 
-  // gfx context interface
   int GetWidth() const;
   int GetHeight() const;
   bool SetViewPort(float fx, float fy , float fwidth, float fheight, bool intersectPrevious = false);
@@ -112,6 +119,9 @@
    */
   void Clear(KODI::UTILS::COLOR::Color color);
   void GetAllowedResolutions(std::vector<RESOLUTION> &res);
+  uint32_t ConvertRenderStereoModeToMode3dFlags(RenderStereoMode stereo_mode) const;
+  RenderStereoMode ConvertMode3dFlagsToRenderStereoMode(uint32_t mode3dFlags) const;
+  RenderStereoMode GetRenderStereoModeFromResolution(RESOLUTION resolution) const;
   /*! \brief Sets the direction of the current rendering pass.
    \param renderOrder direction of the pass
    */
@@ -235,6 +245,7 @@
   bool m_bCalibrating = false;
   RESOLUTION m_Resolution = RES_INVALID;
   float m_fFPSOverride = 0.0f;
+  bool m_dvOutput = false;
 
   RESOLUTION_INFO m_windowResolution;
   std::stack<CPoint> m_cameras;
