diff --git a/xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp b/xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp
index dfd4b31227..1774eba28f 100644
--- a/xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp
+++ b/xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.cpp
@@ -241,7 +241,7 @@ void CRenderer::Render(COverlay* o)
     }
   }
 
-  state.x += GetStereoscopicDepth();
+  state.x += GetStereoscopicDepth(o->m_pgsSubtitle, o->m_3dSubtitleDepth);
 
   o->Render(state);
 }
@@ -418,20 +418,6 @@ std::shared_ptr<COverlay> CRenderer::ConvertLibass(
   rOpts.frameWidth = m_rv.Width();
   rOpts.frameHeight = m_rv.Height();
 
-  // Render subtitle of half-sbs and half-ou video in full screen, not in half screen
-  if (m_stereomode == "left_right" || m_stereomode == "right_left")
-  {
-    // only half-sbs video, sbs video don't need to change source size
-    if (rOpts.sourceWidth / rOpts.sourceHeight < 1.2f)
-      rOpts.sourceWidth = m_rs.Width() * 2;
-  }
-  else if (m_stereomode == "top_bottom" || m_stereomode == "bottom_top")
-  {
-    // only half-ou video, ou video don't need to change source size
-    if (rOpts.sourceWidth / rOpts.sourceHeight > 2.5f)
-      rOpts.sourceHeight = m_rs.Height() * 2;
-  }
-
   // Set position of subtitles based on video calibration settings
   RESOLUTION_INFO resInfo = CServiceBroker::GetWinSystem()->GetGfxContext().GetResInfo();
   // Keep track of subtitle position value change,
@@ -473,12 +459,22 @@ std::shared_ptr<COverlay> CRenderer::ConvertLibass(
     // default position of the text, different from the other alignment positions
     double posPx = static_cast<double>(m_subtitlePosition - resInfo.Overscan.top);
 
+    double frameHeight = static_cast<double>(rOpts.frameHeight);
+
+    if (m_stereomode == "top_bottom" || m_stereomode == "bottom_top")
+    {
+      // only half-ou video, ou video don't need to correct frame height
+      if (rOpts.sourceWidth / rOpts.sourceHeight > 1.2f)
+        frameHeight *= 2.0;
+    }
+
     int assPlayResY = o.GetLibassHandler()->GetPlayResY();
     double assVertMargin = static_cast<double>(overlayStyle->marginVertical) *
                            (static_cast<double>(assPlayResY) / 720);
-    double vertMarginScaled = assVertMargin / assPlayResY * static_cast<double>(rOpts.frameHeight);
 
-    double pos = posPx / (static_cast<double>(rOpts.frameHeight) - vertMarginScaled);
+    double vertMarginScaled = assVertMargin / assPlayResY * frameHeight;
+    double pos = posPx / (frameHeight - vertMarginScaled);
+
     rOpts.position = 100 - pos * 100;
   }
   else if (m_subtitleAlign == SUBTITLES::Align::BOTTOM_OUTSIDE)
@@ -575,6 +571,12 @@ std::shared_ptr<COverlay> CRenderer::Convert(CDVDOverlay& o, double pts)
   else if (o.IsOverlayType(DVDOVERLAY_TYPE_SPU))
     r = COverlay::Create(static_cast<CDVDOverlaySpu&>(o));
 
+  if (r)
+  {
+    r->m_3dSubtitleDepth = o.m_3dSubtitleDepth;
+    r->m_pgsSubtitle = o.IsOverlayType(DVDOVERLAY_TYPE_IMAGE);
+  }
+
   m_textureCache[m_textureid] = r;
   o.m_textureid = m_textureid;
   m_textureid++;
