diff --git a/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp b/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp
index 66a7d39..9c0c8e3 100644
--- a/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp
+++ b/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.cpp
@@ -18,8 +18,12 @@
 #include "settings/lib/SettingDefinitions.h"
 #include "utils/MathUtils.h"
 #include "utils/log.h"
+#include "utils/StringUtils.h"
+#include "utils/SysfsUtils.h"
 #include "windowing/GraphicContext.h"
 #include "windowing/WinSystem.h"
+#include "xbmc/application/Application.h"
+#include "application/ApplicationPlayer.h"
 
 #include <algorithm>
 #include <cstdlib> // std::abs(int) prototype
@@ -126,6 +130,46 @@ void CBaseRenderer::CalcDestRect(float offsetX,
   // and the output pixel ratio setting)
 
   float outputFrameRatio = inputFrameRatio / CServiceBroker::GetWinSystem()->GetGfxContext().GetResInfo().fPixelRatio;
+
+  static int countdown = 0;
+
+  std::string hdmiAspect;
+  SysfsUtils::GetString("/sys/class/amhdmitx/amhdmitx0/aspect", hdmiAspect);
+
+  if (((MathUtils::FloatEquals(inputFrameRatio, 4.0f / 3.0f, 0.01f) &&
+      !MathUtils::FloatEquals(inputFrameRatio, (float) m_sourceWidth / (float) m_sourceHeight, 0.01f)) ||
+      (StringUtils::EndsWith(hdmiAspect, "0x01/0x08") && countdown > 0)) &&
+      CServiceBroker::GetSettingsComponent()->GetSettings()->GetInt(CSettings::SETTING_VIDEOPLAYER_STRETCH43) == ViewModeHDMIAVI &&
+      m_videoSettings.m_ViewMode == ViewModeNormal)
+  {
+    auto& components = CServiceBroker::GetAppComponents();
+    const auto appPlayer = components.GetComponent<CApplicationPlayer>();
+    if (appPlayer->IsPaused())
+      countdown = 20;
+    else if (countdown > 0)
+      countdown--;
+
+    if (countdown > 0 || MathUtils::FloatEquals(inputFrameRatio, 4.0f / 3.0f, 0.01f))
+    {
+      if (!StringUtils::EndsWith(hdmiAspect, "0x01/0x08"))
+      {
+        CLog::Log(LOGDEBUG, "CBaseRenderer::CalcNormalRenderRect Setting aspect to 4:3");
+        SysfsUtils::SetInt("/sys/class/amhdmitx/amhdmitx0/aspect", 1);
+      }
+      outputFrameRatio = ((float) CServiceBroker::GetWinSystem()->GetGfxContext().GetResInfo().iWidth) /
+          ((float) CServiceBroker::GetWinSystem()->GetGfxContext().GetResInfo().iHeight);
+    }
+  }
+  else
+  {
+    if (!StringUtils::EndsWith(hdmiAspect, "0x02/0x08"))
+    {
+      CLog::Log(LOGDEBUG, "CBaseRenderer::CalcNormalRenderRect Setting aspect to 16:9");
+      SysfsUtils::SetInt("/sys/class/amhdmitx/amhdmitx0/aspect", 2);
+    }
+    if (CServiceBroker::GetWinSystem()->GetGfxContext().GetResInfo().iWidth == 720)
+      outputFrameRatio *= 3.0f / 4.0f;
+  }
 
   // allow a certain error to maximize size of render area
   float fCorrection = width / height / outputFrameRatio - 1.0f;
@@ -166,6 +210,14 @@ void CBaseRenderer::CalcDestRect(float offsetX,
       newWidth = std::min(width, height);
       newHeight = newWidth * outputFrameRatio;
     }
+    if (CServiceBroker::GetWinSystem()->GetGfxContext().GetResInfo().iWidth == 720) {
+      if (CServiceBroker::GetWinSystem()->GetGfxContext().GetResInfo().iHeight == 576)
+        newWidth *= (225.0f / 256.0f);
+      if (CServiceBroker::GetWinSystem()->GetGfxContext().GetResInfo().iHeight == 480)
+        newWidth *= (81.0f / 64.0f);
+      if (MathUtils::FloatEquals(inputFrameRatio, 1.7777f, 0.01f))
+        newWidth *= (9.0f / 16.0f);
+    }
   }
 
   // Scale the movie up by set zoom amount
