diff --git a/xbmc/cores/AudioEngine/Utils/AEChannelInfo.cpp b/xbmc/cores/AudioEngine/Utils/AEChannelInfo.cpp
index e3d74ca3a1..d724ab4bd5 100644
--- a/xbmc/cores/AudioEngine/Utils/AEChannelInfo.cpp
+++ b/xbmc/cores/AudioEngine/Utils/AEChannelInfo.cpp
@@ -313,7 +313,8 @@ bool CAEChannelInfo::ContainsChannels(const CAEChannelInfo& rhs) const
 {
   for (unsigned int i = 0; i < rhs.m_channelCount; ++i)
   {
-    if (!HasChannel(rhs.m_channels[i]))
+   if (rhs.m_channels[i] > AE_CH_NULL && rhs.m_channels[i] < AE_CH_UNKNOWN1 &&
+        !HasChannel(rhs.m_channels[i]))
       return false;
   }
   return true;
