diff --git a/xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp b/xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp
index 7bcdf8c684..a3ddcb7ccd 100644
--- a/xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp
+++ b/xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp
@@ -279,7 +279,8 @@ ASS_Image* CDVDSubtitlesLibass::RenderImage(double pts,
   // Reversed par value
   // from: >1 tighter pixels, <1 wider pixels
   // to: <1 tighter pixels, >1 wider pixels
-  float par = (opts.m_par - 2.0f) * -1;
+  float par = 1.0f / opts.m_par;
+
   ass_set_pixel_aspect(m_renderer, static_cast<double>(par));
 
   ass_set_frame_size(m_renderer, static_cast<int>(opts.frameWidth),
