diff --git a/xbmc/settings/DisplaySettings.h b/xbmc/settings/DisplaySettings.h
index 9f4be62..34e1fab 100644
--- a/xbmc/settings/DisplaySettings.h
+++ b/xbmc/settings/DisplaySettings.h
@@ -79,6 +79,7 @@ public:
   {
     return GetStringFromResolution(resolution, refreshrate);
   }
+  std::string GetDisplayStringFromRes(const RESOLUTION resolution, bool hiDpi = false, bool withRefreshRate = true) const;
 
   void ApplyCalibrations();
   void UpdateCalibrations();
